File: backend/app/api/maintenance.py
import asyncio
import logging
import uuid
from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app import storage
from app.api.deps import get_session, require_auth
from app.models import DropFile, DropItem
from app.schemas import (
    MissingFileItem,
    OrphanFileItem,
    StorageCheckResponse,
    StorageFixResponse,
)
from app.ws import manager

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/maintenance", tags=["maintenance"])


@router.get("/storage-check", response_model=StorageCheckResponse)
async def check_storage(
    session: AsyncSession = Depends(get_session),
    _: tuple = Depends(require_auth),
) -> StorageCheckResponse:
    # 1. Fetch all items with files from database (including soft-deleted)
    result = await session.execute(
        select(DropItem).options(selectinload(DropItem.files))
    )
    items = list(result.scalars().all())

    db_item_count = len(items)
    db_file_count = 0
    all_db_shas: set[str] = set()
    missing_files: list[MissingFileItem] = []

    for item in items:
        for f in item.files:
            db_file_count += 1
            sha = f.sha256.lower()
            all_db_shas.add(sha)
            if f.uploaded_at is not None:
                if not storage.file_exists(sha):
                    missing_files.append(
                        MissingFileItem(
                            item_id=item.id,
                            item_kind=item.kind,
                            item_note=item.note,
                            item_created_at=item.created_at,
                            item_is_ephemeral=item.is_ephemeral,
                            item_is_secret=item.is_secret,
                            item_deleted_at=item.deleted_at,
                            file_id=f.id,
                            file_name=f.file_name,
                            file_size=f.size,
                            sha256=sha,
                        )
                    )

    # 2. Scan physical files on disk
    disk_files = await asyncio.to_thread(storage.scan_physical_files)
    total_disk_files = len(disk_files)
    total_disk_size = sum(f["size"] for f in disk_files)

    orphan_files: list[OrphanFileItem] = []
    for file_info in disk_files:
        sha = file_info["sha256"]
        if sha not in all_db_shas:
            orphan_files.append(
                OrphanFileItem(
                    sha256=sha,
                    size=file_info["size"],
                    path=file_info["path"],
                )
            )

    status = "healthy" if not missing_files and not orphan_files else "issues_found"

    return StorageCheckResponse(
        status=status,
        total_db_items=db_item_count,
        total_db_files=db_file_count,
        total_disk_files=total_disk_files,
        total_disk_size=total_disk_size,
        missing_files=missing_files,
        orphan_files=orphan_files,
    )


@router.post("/storage-fix", response_model=StorageFixResponse)
async def fix_storage(
    session: AsyncSession = Depends(get_session),
    _: tuple = Depends(require_auth),
) -> StorageFixResponse:
    # 1. Fetch all items from DB
    result = await session.execute(
        select(DropItem).options(selectinload(DropItem.files))
    )
    items = list(result.scalars().all())

    all_db_shas: set[str] = set()
    broken_items: list[DropItem] = []

    for item in items:
        item_is_broken = False
        for f in item.files:
            sha = f.sha256.lower()
            all_db_shas.add(sha)
            if f.uploaded_at is not None and not storage.file_exists(sha):
                item_is_broken = True

        if item_is_broken:
            broken_items.append(item)

    # 2. Clean orphan physical files on disk
    disk_files = await asyncio.to_thread(storage.scan_physical_files)
    deleted_orphan_count = 0
    deleted_orphan_size = 0

    for file_info in disk_files:
        sha = file_info["sha256"]
        if sha not in all_db_shas:
            if storage.remove_orphan_file(file_info["full_path"]):
                deleted_orphan_count += 1
                deleted_orphan_size += file_info["size"]

    # 3. Clean broken DB items
    deleted_broken_count = 0
    other_shas_to_check: set[str] = set()

    for item in broken_items:
        item_id = item.id
        for f in item.files:
            other_shas_to_check.add(f.sha256.lower())
        await session.delete(item)
        deleted_broken_count += 1
        await manager.broadcast({"type": "item_deleted", "id": str(item_id)})

    if deleted_broken_count > 0:
        await session.commit()
        # Clean up any physical files from the deleted broken items if unreferenced
        for sha in other_shas_to_check:
            await storage.delete_file_if_unreferenced(sha, session)

    logger.info(
        "storage fix completed: deleted %d orphan files (%d bytes), deleted %d broken items",
        deleted_orphan_count,
        deleted_orphan_size,
        deleted_broken_count,
    )

    return StorageFixResponse(
        deleted_orphan_files_count=deleted_orphan_count,
        deleted_orphan_files_size=deleted_orphan_size,
        deleted_broken_items_count=deleted_broken_count,
    )

File: backend/internal/api/maintenance.go
package api

import (
	"context"
	"database/sql"
	"net/http"
	"strings"
	"time"

	"github.com/google/uuid"
	"privatedrop/internal/database"
	"privatedrop/internal/models"
	"privatedrop/internal/storage"
	"privatedrop/internal/ws"
)

type MaintenanceHandler struct {
	db        *database.DB
	storage   *storage.StorageManager
	wsManager *ws.ConnectionManager
}

func NewMaintenanceHandler(
	db *database.DB,
	storage *storage.StorageManager,
	wsManager *ws.ConnectionManager,
) *MaintenanceHandler {
	return &MaintenanceHandler{
		db:        db,
		storage:   storage,
		wsManager: wsManager,
	}
}

type dbItemData struct {
	ID          uuid.UUID
	Kind        string
	Note        *string
	IsEphemeral bool
	IsSecret    bool
	ExpiresAt   *time.Time
	DeletedAt   *time.Time
	CreatedAt   time.Time
	Files       []dbFileData
}

type dbFileData struct {
	ID         uuid.UUID
	FileName   string
	MimeType   string
	Size       int64
	SHA256     string
	UploadedAt *time.Time
}

func (h *MaintenanceHandler) fetchAllItems(ctx context.Context) ([]dbItemData, error) {
	query := `SELECT id, kind, note, is_ephemeral, is_secret, expires_at, deleted_at, created_at FROM drop_items`
	rows, err := h.db.QueryContext(ctx, query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var items []dbItemData
	itemMap := make(map[uuid.UUID]int)

	for rows.Next() {
		var idStr, kind string
		var note sql.NullString
		var isEphemeral, isSecret bool
		var expiresAt, deletedAt sql.NullTime
		var createdAt time.Time

		if err := rows.Scan(&idStr, &kind, &note, &isEphemeral, &isSecret, &expiresAt, &deletedAt, &createdAt); err == nil {
			id, _ := uuid.Parse(idStr)
			item := dbItemData{
				ID:          id,
				Kind:        kind,
				IsEphemeral: isEphemeral,
				IsSecret:    isSecret,
				CreatedAt:   createdAt,
			}
			if note.Valid {
				item.Note = &note.String
			}
			if expiresAt.Valid {
				t := expiresAt.Time.UTC()
				item.ExpiresAt = &t
			}
			if deletedAt.Valid {
				t := deletedAt.Time.UTC()
				item.DeletedAt = &t
			}
			itemMap[id] = len(items)
			items = append(items, item)
		}
	}
	rows.Close()

	// Load files
	fQuery := `SELECT id, item_id, file_name, mime_type, size, sha256, uploaded_at FROM drop_files`
	fRows, err := h.db.QueryContext(ctx, fQuery)
	if err == nil {
		defer fRows.Close()
		for fRows.Next() {
			var fIDStr, itemIDStr, fName, mimeType, sha string
			var size int64
			var uploadedAt sql.NullTime

			if err := fRows.Scan(&fIDStr, &itemIDStr, &fName, &mimeType, &size, &sha, &uploadedAt); err == nil {
				itemID, _ := uuid.Parse(itemIDStr)
				fID, _ := uuid.Parse(fIDStr)
				file := dbFileData{
					ID:       fID,
					FileName: fName,
					MimeType: mimeType,
					Size:     size,
					SHA256:   strings.ToLower(sha),
				}
				if uploadedAt.Valid {
					t := uploadedAt.Time.UTC()
					file.UploadedAt = &t
				}
				if idx, ok := itemMap[itemID]; ok {
					items[idx].Files = append(items[idx].Files, file)
				}
			}
		}
	}

	return items, nil
}

func (h *MaintenanceHandler) StorageCheck(w http.ResponseWriter, r *http.Request) {
	items, err := h.fetchAllItems(r.Context())
	if err != nil {
		WriteError(w, http.StatusInternalServerError, "failed to check storage")
		return
	}

	dbItemCount := len(items)
	dbFileCount := 0
	allDBSHAs := make(map[string]bool)
	var missingFiles []models.MissingFileItem

	for _, item := range items {
		for _, f := range item.Files {
			dbFileCount++
			allDBSHAs[f.SHA256] = true
			if f.UploadedAt != nil {
				if !h.storage.FileExists(f.SHA256) {
					missingFiles = append(missingFiles, models.MissingFileItem{
						ItemID:          item.ID,
						ItemKind:        item.Kind,
						ItemNote:        item.Note,
						ItemCreatedAt:   item.CreatedAt,
						ItemIsEphemeral: item.IsEphemeral,
						ItemIsSecret:    item.IsSecret,
						ItemDeletedAt:   item.DeletedAt,
						FileID:          f.ID,
						FileName:        f.FileName,
						FileSize:        f.Size,
						SHA256:          f.SHA256,
					})
				}
			}
		}
	}

	diskFiles, err := h.storage.ScanPhysicalFiles()
	if err != nil {
		WriteError(w, http.StatusInternalServerError, "failed to scan physical files")
		return
	}

	totalDiskFiles := len(diskFiles)
	var totalDiskSize int64
	var orphanFiles []models.OrphanFileItem

	for _, df := range diskFiles {
		totalDiskSize += df.Size
		if !allDBSHAs[df.SHA256] {
			orphanFiles = append(orphanFiles, models.OrphanFileItem{
				SHA256: df.SHA256,
				Size:   df.Size,
				Path:   df.Path,
			})
		}
	}

	status := "healthy"
	if len(missingFiles) > 0 || len(orphanFiles) > 0 {
		status = "issues_found"
	}

	if missingFiles == nil {
		missingFiles = []models.MissingFileItem{}
	}
	if orphanFiles == nil {
		orphanFiles = []models.OrphanFileItem{}
	}

	WriteJSON(w, http.StatusOK, models.StorageCheckResponse{
		Status:         status,
		TotalDBItems:   dbItemCount,
		TotalDBFiles:   dbFileCount,
		TotalDiskFiles: totalDiskFiles,
		TotalDiskSize:  totalDiskSize,
		MissingFiles:   missingFiles,
		OrphanFiles:    orphanFiles,
	})
}

func (h *MaintenanceHandler) StorageFix(w http.ResponseWriter, r *http.Request) {
	items, err := h.fetchAllItems(r.Context())
	if err != nil {
		WriteError(w, http.StatusInternalServerError, "failed to fetch items for fix")
		return
	}

	allDBSHAs := make(map[string]bool)
	var brokenItems []dbItemData

	for _, item := range items {
		itemIsBroken := false
		for _, f := range item.Files {
			allDBSHAs[f.SHA256] = true
			if f.UploadedAt != nil && !h.storage.FileExists(f.SHA256) {
				itemIsBroken = true
			}
		}
		if itemIsBroken {
			brokenItems = append(brokenItems, item)
		}
	}

	// 1. Delete orphan physical files on disk
	diskFiles, err := h.storage.ScanPhysicalFiles()
	if err != nil {
		WriteError(w, http.StatusInternalServerError, "failed to scan physical files")
		return
	}

	deletedOrphanCount := 0
	var deletedOrphanSize int64

	for _, df := range diskFiles {
		if !allDBSHAs[df.SHA256] {
			if h.storage.RemoveOrphanFile(df.FullPath) {
				deletedOrphanCount++
				deletedOrphanSize += df.Size
			}
		}
	}

	// 2. Clean broken DB items
	deletedBrokenCount := 0
	otherSHAsToCheck := make(map[string]bool)

	for _, item := range brokenItems {
		for _, f := range item.Files {
			otherSHAsToCheck[f.SHA256] = true
		}

		delQuery := "DELETE FROM drop_items WHERE id = $1"
		_, err := h.db.ExecContext(r.Context(), delQuery, item.ID.String())
		if err != nil {
			continue
		}
		deletedBrokenCount++
		h.wsManager.Broadcast(map[string]interface{}{
			"type": "item_deleted",
			"id":   item.ID.String(),
		})
	}

	for sha := range otherSHAsToCheck {
		_, _ = h.storage.DeleteFileIfUnreferenced(r.Context(), h.db.DB, sha)
	}

	WriteJSON(w, http.StatusOK, models.StorageFixResponse{
		DeletedOrphanFilesCount: deletedOrphanCount,
		DeletedOrphanFilesSize:  deletedOrphanSize,
		DeletedBrokenItemsCount: deletedBrokenCount,
	})
}

